src/environments.js

```diff
@@ -1,18 +1,96 @@
-/* eslint no-constant-condition:0 */
+// @flow
 import ParseNode from "./ParseNode";
 import ParseError from "./ParseError";
 
+import type Parser from "./Parser";
+import type {ArgType, Mode, StyleStr} from "./types";
+
+/**
+ * The context contains the following properties:
+ *  - mode: current parsing mode.
+ *  - envName: the name of the environment, one of the listed names.
+ *  - parser: the parser object.
+ *  - positions: the positions associated with these arguments from args.
+ */
+type EnvContext = {
+  mode: Mode,
+  envName: string,
+  parser: Parser,
+  positions: number[],
+};
+
+/**
+ * List of ParseNodes followed by an array of positions.
+ * Returned by `Parser`'s `parseArguments`.
+ */
+type ParseResult = (ParseNode | number[])[] | ParseNode;
```

Comment:
When do we return a result with type `number[][]`?

Reply:
`Parser`'s `parseArgument` always returns the list of positions as the last element.
So it returns something of the format `[ParseNode, ParseNode, ..., ParseNode, number[]]`. So this happens any time you see a `defineFunction` with code like
```javascript
    res = parseArray(context.parser, res, "display");
    return res;
```
We need to refactor this at some point, but I'd rather do it right before or during `ParseNode` being ported. It would help to have maximal support from type-safety while doing this since `ParseResult` is simply stored here and used various places (including possible buildHtml, but I haven't read that part of the code base yet).

Reply:
I see, it's documented in the comment above.

> We need to refactor this at some point, but I'd rather do it right before or during ParseNode being ported.

Sounds good.